controllers/capa_utils.go

```diff
@@ -41,3 +51,46 @@ func isRoleUsedElsewhere(ctx context.Context, ctrlClient client.Client, roleName
 
 	return false, err
 }
+
+func removeFinalizer(ctx context.Context, k8sClient client.Client, object client.Object, role string) error {
+	logger := log.FromContext(ctx)
+
+	if !controllerutil.ContainsFinalizer(object, key.FinalizerName(role)) {
+		logger.Info("finalizer already removed")
+		return nil
+	}
+
+	for i := 1; i <= maxPatchRetries; i++ {
+		patchHelper, err := patch.NewHelper(object, k8sClient)
+		if err != nil {
+			logger.Error(err, "failed to create patch helper")
+			return errors.WithStack(err)
+		}
+		controllerutil.RemoveFinalizer(object, key.FinalizerName(role))
+		err = patchHelper.Patch(ctx, object)
+
+		// If another controller has removed its finalizer while we're
+		// reconciling this will fail with "Forbidden: no new finalizers can be
+		// added if the object is being deleted". The actual response code is
+		// 422 Unprocessable entity, which maps to StatusReasonInvalid in the
+		// k8serrors package. We have to get the cluster again with the now
+		// removed finalizer(s) and try again.
+		invalidErr := errutils.FilterOut(err, func(err error) bool {
+			return !k8serrors.IsInvalid(err)
+		})
+
+		if invalidErr != nil && i < maxPatchRetries {
+			logger.Info(fmt.Sprintf("patching object failed, trying again: %s", err.Error()))
+			if err := k8sClient.Get(ctx, client.ObjectKeyFromObject(object), object); err != nil {
+				return microerror.Mask(err)
+			}
+			continue
+		}
+		if err != nil {
+			logger.Error(err, "failed to remove finalizers")
+			return microerror.Mask(err)
+		}
```

Comment:
Isn't this missing a `return`/`break` in case of success? I mean checking `if err == nil { break }` directly after `patchHelper.Patch(...)`.

Reply:
Yeah you're right. I've changed it, can you take a look?